libnetwork/idm/idm.go

```diff
@@ -4,40 +4,39 @@ package idm
 import (
 	"errors"
 	"fmt"
+	"sync"
 
-	"github.com/docker/docker/libnetwork/bitseq"
-	"github.com/docker/docker/libnetwork/datastore"
+	"github.com/docker/docker/libnetwork/bitmap"
 )
 
 // Idm manages the reservation/release of numerical ids from a contiguous set
 type Idm struct {
 	start  uint64
 	end    uint64
-	handle *bitseq.Handle
+	mu     sync.Mutex
+	handle *bitmap.Bitmap
 }
 
+// placeholder is a type for function arguments which need to be present for Swarmkit
+// to compile, but for which the only acceptable value is nil.
+type placeholder *struct{}
```

Comment:
I'm sure there's a train of thought here why to use a `*struct{}` 🤔 

I was wondering; would it work to change this to `any`, and change `New` to have two placeholders;

```go
func New(_, _ placeholder, start, end uint64) (*Idm, error) {
```

We can also start passing `nil` for both in tests and other places, which makes it very clear "nope, not used";

<img width="779" alt="Screenshot 2023-07-05 at 18 51 31" src="https://github.com/moby/moby/assets/1804568/8a0cdab7-cd35-4104-8fa0-73a6c54a421d">

Reply:
If that works; here's my changes committed; https://github.com/corhere/moby/compare/libn/drop-bitseq...thaJeztah:docker:TEMP_review_45883?expand=1

Reply:
This is literally the same pattern I used in package drvregistry! The point of `placeholder` is that it is a [unit type](https://en.wikipedia.org/wiki/Unit_type) in which it is a compile error to pass any value aside from nil. Changing it to `any` would make it a [top type](https://en.wikipedia.org/wiki/Top_type) with no compile-time assertion if you're holding it wrong.

The ID (string) argument is only meaningful when using a datastore-backed idm and is not otherwise exposed in the API so there is no way to observe whether or not the argument is being used for anything. So I don't think it's a big deal to leave strings in the tests.

> We can also start passing `nil` for both in tests and other places, which makes it very clear "nope, not used";

After #45882 there will be no other places aside from Swarmkit. Then [I want to move the idm package into Swarmkit where we can then drop those arguments outright.](https://github.com/moby/swarmkit/pull/3139#discussion_r1252393157)